packages/cardpay-reward-programs/cardpay_reward_programs/rule.py

```diff
@@ -0,0 +1,90 @@
+import hashlib
+import json
+from abc import ABC, abstractmethod
+
+import duckdb
+import pandas as pd
+from cardpay_reward_programs.utils import get_files
+
+
+class Rule(ABC):
+    """
+    A single image should run only a single rule
+    """
+
+    def __init__(self, core_parameters, user_defined_parameters):
+        self.set_core_parmaters(**core_parameters)
+        self.set_user_defined_parameters(**user_defined_parameters)
+
+    def set_core_parmaters(
+        self, subgraph_config_location, payment_cycle_length, valid_from, valid_to, token
+    ):
+        self.subgraph_config_location = subgraph_config_location
+        self.payment_cycle_length = payment_cycle_length
+        self.valid_from = valid_from
+        self.valid_to = valid_to
+        self.token = token
+
+    def get_core_hash(self):
+        core_parameters = {
+            "subgraph_config_location": self.subgraph_config_location,
+            "payment_cycle_length": self.payment_cycle_length,
+            "valid_from": self.valid_from,
+            "valid_to": self.valid_to,
+            "token": self.token,
+        }
+        o = json.dumps(core_parameters, sort_keys=True)
+        return hashlib.md5(o.encode("utf-8")).hexdigest()
```

Comment:
The hashing here is to get a unique value that represents the config for the rules right? Is there a reason to split it into core & user defined? We could just hash the params being passed in in total.

Reply:
I agree with you. I was really wanting to make the hash a staticmethod type. But there is a few refactoring questions around it. 

Question: should the `core_params` and `user_defined_params` be variables that are set only upon instantiating of the class? Think the implications around it is around how an analytical miner interacts with the docker container. If a rule is updated with params, I presume the docker container will be restarted. Available for call to clarify this.

Reply:
Not sure if this was before/after we talked the other day, so this may have been clarified already but

* Can set them on instantiating the class - neither will be built into the image
* I think it's reasonable that once a miner starts working on something they don't need to check for rule changes
* * That is - if a calculation can be triggered (progam not locked, data available and the right payment cycle/etc), it's fair game.
* * My expectation is that most calculations will be fast - these aggregations & duckdb based queries are (once data is downloaded) sub-second pieces of work so the window for timing issues around changing rules is very small
* Each piece of work is independent - the containers will be started, the task run, then cleared away rather than having long running containers. Miners would have their own scheduler/runner (to look for things that can be run, perhaps work out the most valuable ones, and then trigger the tasks) which would be long running